### DISCUSSION TITLE
How to do allow on two decimal points on a number input textbox

### DISCUSSION THREAD

**Discussion by Cants820** [react-hook-form/react-hook-form#8210] [Q&A]
Hello,

I have been struggling for days to solve this issue. wherein you only allow 2 decimal points on an input box. Here is the code what i am working at:  

Any inputs would be greatly appreciated

```
const handleChange = e => {
        const { name, value } = e.target;
   
           // check if value includes a decimal point
           if (value.match(/\./g)) {
            const [, decimal] = value.split('.');
            
            // restrict value to only 2 decimal places
            if (decimal?.length > 2) {
                // do nothing
                return;
            }
        }

    }
```

Here is my input textbox 
                         ` <Input type="number" name="classNo" onChange={handleChange} defaultValue={classNo} min={'0'} label="Class" valueAsNumber={true} register={register} disabled={viewOnly} />`

Thank you,

**Comment by Moshyfawn** [comment]
You want to use **a controlled** input _(`Controller`)_ instead of an uncontrolled one _(`register`)_ to manipulate the value format.

Uncontrolled input's source of truth for their value is an actual input. So, whatever manipulations you do with the value `onChange` won't propagate to the input's value.

With controlled inputs, you store the value outside the input element _(RHF form values state in your case)_ and pass it back down to the input. This way all the value transformations you do will be reflected in the input's value.

```jsx
<Controller
  name="number"
  control={control}
  render={({ field: { onChange, ...fieldProps } }) => (
    <section>
      <label htmlFor="controlled">Controlled</label>
      <input
        {...fieldProps}
        id="controlled"
        onChange={(event) => onChange(limitDecimalPlaces(2)(event))}
        type="number"
        step=".01"
      />
    </section>
  )}
/>
```

Here's [an example CSB](https://codesandbox.io/s/runtime-monad-jusg0p?file=%2Fsrc%2FDemo.tsx)
